src/eins-network-id.c

```diff
@@ -0,0 +1,260 @@
+  g_assert (arp_rx != NULL);
+
+  g_autoptr(GMatchInfo) arp_mi = NULL;
+  if (!g_regex_match (arp_rx, arp, 0, &arp_mi))
+    {
+      g_debug ("couldn’t find HW address in " PROC_IPV4_ARP_FILE);
+      return NULL;
+    }
+
+  const gchar *hw = g_match_info_fetch_named (arp_mi, "HW");
+  g_assert (hw != NULL);
+  return g_ascii_strdown (hw, -1);
+}
+
+#define PROC_IPV6_ROUTE_FILE "/proc/net/ipv6_route"
+/* we want dest/prefix of 0/, any src/prefix, grab the gateway, any metric/ref/use, flags ending 3 = UP + GATEWAY */
+/* <dest 32> <prefix 2> <src 32> <prefix 2> <gateway 32> <metric 8> <refcount 8> <use? 8> <flags 8> <if> */
+#define IPV6_ROUTE_REGEXP "^0{32}\\s00\\s[0-9A-Fa-f]{32}\\s[0-9A-Fa-f]{2}\\s(?P<Gateway>[0-9A-Fa-f]{32})\\s([0-9A-Fa-f]{8}\\s){3}[0-9A-Fa-f]{7}3"
+
+/*
+ * greps the /proc/net/ipv6_route file for the first IPv6 default route. same
+ * assumptions as above apply.
+ */
+static gchar *
+get_ipv6_default_gateway (void)
+{
+  g_autofree gchar *route = NULL;
+  g_autoptr(GError) err = NULL;
+
+  if (!g_file_get_contents (PROC_IPV6_ROUTE_FILE, &route, NULL, &err))
+    {
+      g_debug ("could not load " PROC_IPV6_ROUTE_FILE ": %s", err->message);
+      return NULL;
+    }
+
+  g_autoptr(GRegex) route_rx = g_regex_new (IPV6_ROUTE_REGEXP, G_REGEX_MULTILINE, 0, NULL);
+  g_assert (route_rx != NULL);
+
+  g_autoptr(GMatchInfo) route_mi = NULL;
+  if (!g_regex_match (route_rx, route, 0, &route_mi))
+    {
+      g_debug ("couldn’t match IPv6 default gateway in " PROC_IPV6_ROUTE_FILE);
+      return NULL;
+    }
+
+  const gchar *gateway = g_match_info_fetch_named (route_mi, "Gateway");
+  g_assert (gateway != NULL);
+
+  /* the input to this sscanf call is validated by the regexp above */
+  guint32 addr[8];
+  if (sscanf(gateway,"%4x%4x%4x%4x%4x%4x%4x%4x",
+             &addr[0], &addr[1], &addr[2], &addr[3],
+             &addr[4], &addr[5], &addr[6], &addr[7]) != 8)
+    {
+      g_assert_not_reached ();
+      return NULL;
+    }
+
+  return g_strdup_printf ("%x:%x:%x:%x:%x:%x:%x:%x",
+                          addr[0], addr[1], addr[2], addr[3],
+                          addr[4], addr[5], addr[6], addr[7]);
+}
+
+#define IPV6_NDISC_REGEXP "^(\\S+\\s+){3}lladdr\\s(?P<HW>[0-9A-Fa-f:]{17})"
+
+/*
+ * no IPv6 equivalent of /proc/net/arp exists to view the neighbour
+ * discovery (ndisc) cache, so rather than breaking out netlink, we
+ * invoke the "ip" command for sanity’s sake.
+ */
+static gchar *
+get_ipv6_hwaddr (const gchar *address)
+{
+  const gchar *ndisc_cmd[] = { "ip", "-6", "neigh", "show", address, NULL };
+  g_autofree gchar *ndisc = NULL;
+  g_autofree gchar *ndisc_err = NULL;
+  g_autoptr(GError) err = NULL;
+  gint ret = 0;
+
+  if (!g_spawn_sync (NULL, (gchar **) &ndisc_cmd[0], NULL,
+                     G_SPAWN_SEARCH_PATH, NULL, NULL,
+                     &ndisc, &ndisc_err, &ret, &err) || ret)
+    {
+      g_debug ("could not execute ip -6 neigh command (ret %i): %s", ret,
```

Comment:
Would be good to format `ndisc_cmd` and print it in this debug message properly, so the `address` is in the output and it’s completely reproducible from a terminal.